### PR TITLE
Performance test details relevant to the CC V1 release

### LINKED ISSUE #2478: Performance test for Choreo Connect V 1.0.0
### Description:
Currently there's no details relevant to the Choreo Connect (CC) performance.
Therefore we can arrange a performance test to collect matrices relevant to the performance of Choreo Connect. 

### Steps to reproduce:
N/A

### How will you implement it

- [x] Arrange a Choreo Connect deployment in a cloud environment with a suitable backend (Netty backend).
- [x] Perform the test considering different payload sizes and different concurrent user counts.
- [x] Collect parameters relevant to the CC performance and CC deployment.
- [x] Create a document to demonstrate CC performance test details.

---

### PR DESCRIPTION
**Pull request by slahirucd7**
### Purpose
Adds choreo connect performance test benchmark details to the repository.
diagram

inteceptorDiagram

Links to the Test Result Files:

> https://github.com/wso2/product-microgateway/blob/ec8a360be8c3f391870ffce3953b31e2db83f365/performance/benchmarks/cpus-1/summary.md

> https://github.com/wso2/product-microgateway/blob/ec8a360be8c3f391870ffce3953b31e2db83f365/performance/benchmarks/cpus-2/summary.md

> https://github.com/wso2/product-microgateway/blob/ec8a360be8c3f391870ffce3953b31e2db83f365/performance/benchmarks/inteceptor/summary.md

Resource Files: 
[PerfTestResources.zip](https://github.com/wso2/product-microgateway/files/7585726/PerfTestResources.zip)

### Issues

Fixes #2478
Fixes https://github.com/wso2/product-microgateway/issues/2433

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments
AWS EKS cluster
AWS EC2 instances

| Name                                         | EC2 Instance Type | vCPU | Mem(GiB) |
| ----------------------------- | -----------------    | ---- | -------- |
| Apache JMeter Client          | c5.large          | 2    | 4        |
| Apache JMeter Server 01       | c5.xlarge         | 4    | 8        |
| Apache JMeter Server 02       | c5.xlarge         | 4    | 8        |
| AWS EKS cluster (three nodes) | c5.xlarge         | 4    | 8        |

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: performance/benchmarks/cpus-1/summary.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+Below diagram shows the test setup.
+![picture](images/diagram.png)
+
+| Name                          | EC2 Instance Type | vCPU | Mem(GiB) |
+| ----------------------------- | ----------------- | ---- | -------- |
+| Apache JMeter Client          | c5.large          | 2    | 4        |
+| Apache JMeter Server 01       | c5.xlarge         | 4    | 8        |
+| Apache JMeter Server 02       | c5.xlarge         | 4    | 8        |
+| AWS EKS cluster (three nodes) | c5.xlarge         | 4    | 8        |
```
Comment (pubudu538) [inline performance/benchmarks/cpus-1/summary.md:24]: Lets include specs of router, adapter, enforcer, netty, and nginx.

File: performance/benchmarks/cpus-1/summary.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+We executed tests for different numbers of concurrent users and message sizes (payloads).
+
+The main performance metrics:
+
+1. **Throughput**: The number of requests that the WSO2 API Microgateway processes during a specific time interval (e.g. per second).
```
Comment (pubudu538) [inline performance/benchmarks/cpus-1/summary.md:30]: s/WSO2 API Microgateway/Choreo Connect

File: performance/benchmarks/cpus-1/summary.md (hunk 1 above)
Comment (pubudu538) [reply performance/benchmarks/cpus-1/summary.md:24]: Lets include these after line 48.

Comment (slahirucd7) [reply performance/benchmarks/cpus-1/summary.md:24]: Fixed with: 3e6d38b

File: performance/benchmarks/cpus-1/summary.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+[AWS EKS cluster with **c5.xlarge** Amazon EC2 instances](https://aws.amazon.com/eks/?nc2=type_a) were used to deploy WSO2 Choreo Connect.
+When doing below test scenarios, `--cpus` option is provided as 1 and concurrency level for the router provided as 2. Below table includes
+configuration details relevant to the Choreo Connect deployment.
+
+|Container Name|Requesting Memory Amount (Mi)|Requesting CPU Amount (m)|Limiting Memory Amount (Mi)|Limiting CPU Amount (m)|
+|--------------|-----------------------------|-------------------------|---------------------------|-----------------------|
+|Adapter       |500                          |500                      |500                        |500                    |
+|Enforcer      |1000                         |1000                     |1000                       |1000                   |
+|Router        |500                          |1000                     |500                        |1000                   |
```
Comment (pubudu538) [inline performance/benchmarks/cpus-1/summary.md:56]: Lets include backend details as well

Comment (slahirucd7) [reply performance/benchmarks/cpus-1/summary.md:56]: Fixed with: https://github.com/wso2/product-microgateway/pull/2462/commits/2da739bf7e2640cd284bde832c9340c5f2709d82

**Review by pubudu538 — APPROVED** [review]

File: performance/benchmarks/inteceptor/summary.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+In addition to the above metrics, we measure the load average and several other API request related metrics.
+
+The following are the test parameters.
+
+| Test Parameter       | Description                                                     | Values                      |
+| -------------------- | --------------------------------------------------------------- | --------------------------- |
+| Scenario Name        | The name of the test scenario.                                  | Refer to the above table.   |
+| Heap Size            | The amount of memory allocated to the application               | 1024M                       |
+| Concurrent Users     | The number of users accessing the application at the same time. | 10, 50, 100, 200, 500, 1000 |
```
Comment (renuka-fernando) [inline performance/benchmarks/inteceptor/summary.md:41]: For interceptor scenario is it only 100

File: performance/benchmarks/inteceptor/summary.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+| Message Size (Bytes) | The request payload size in Bytes.                              | 50, 1024, 10240, 102400     |
```
Comment (renuka-fernando) [inline performance/benchmarks/inteceptor/summary.md:42]: 1024 only

File: performance/benchmarks/inteceptor/summary.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+When doing below test scenarios, `--cpus` option is provided as 1 and concurrency level for the router provided as 2. Below table includes
+configuration details relevant to the Choreo Connect deployment.
+
+|Container Name|Requesting Memory Amount (Mi)|Requesting CPU Amount (m)|Limiting Memory Amount (Mi)|Limiting CPU Amount (m)|
+|--------------|-----------------------------|-------------------------|---------------------------|-----------------------|
+|Adapter       |500                          |500                      |500                        |500                    |
+|Enforcer      |1000                         |1000                     |1000                       |1000                   |
+|Router        |500                          |1000                     |500                        |1000                   |
+|Netty backend |4096                         |2000                     |6114                       |2000                   |
+
```
Comment (renuka-fernando) [inline performance/benchmarks/inteceptor/summary.md:58]: @pubudu538 shall we add the interceptor container as well?
- Interceptor
- replicas: 2
- resources
    ```yaml
    requests:
      memory: "500Mi"
      cpu: "1000m"
    limits:
      memory: "500Mi"
      cpu: "1000m"
    ```

Comment (pubudu538) [reply performance/benchmarks/inteceptor/summary.md:58]: +1

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] dev-deployment : 20211215.32

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20211215.43

**Pull request merged**